### PR TITLE
Add compatibility checks for C++ extensions

### PR DESCRIPTION
**Pull request by fmassa**
Fixes https://github.com/pytorch/vision/issues/2148

For now, I don't print the PyTorch and torchvision in the error messages as it would require a bit more work (the torchvision version is a global variable if just imported from `.version`), but I believe it's already an improvement compared to what we had before.

Another thing we might want to consider doing is to add the original stack trace in the error message, but again if we want to make it work with torchscript we would probably need to wrap this on a function that returns the string.

@eellison do you see any issue with the current approach of having the `_has_ops` implementation be switched at runtime?

### CODE REVIEW

File: torchvision/extension.py [hunk 1]
```diff
@@ -1,6 +1,10 @@
 _HAS_OPS = False
```
Comment (fmassa) [inline torchvision/extension.py:1]: This can probably be removed now

File: torchvision/extension.py [hunk 2]
```diff
@@ -23,10 +27,25 @@ def _register_extensions():
 try:
     _register_extensions()
     _HAS_OPS = True
+    def _has_ops():
+        return True
 except (ImportError, OSError):
     pass
 
 
+def _assert_has_ops():
```
Comment (fmassa) [inline torchvision/extension.py:36]: I should do something similar for the video and image ops, but I'll do that in a follow-up PR once we define more precisely if this implementation is enough

File: torchvision/extension.py [hunk 3]
```diff
@@ -23,10 +27,25 @@ def _register_extensions():
 try:
     _register_extensions()
     _HAS_OPS = True
+    def _has_ops():
+        return True
 except (ImportError, OSError):
     pass
```
Comment (ppwwyyxx) [inline torchvision/extension.py:33]: Thanks for addressing this!
Perhaps it's possible to define `_has_ops` function here that returns `(bool has_ops, str err_msg)`. Then the exact error message from the exception can be used in `_assert_has_ops`

**Comment by cpuhrsch** [comment]
Since torchvision is using TORCH_LIBRARY we should probably also bring this to attention of @jamesr66a for torchbind.

File: torchvision/extension.py (hunk 3 above)
Comment (fmassa) [reply torchvision/extension.py:33]: Thanks for the review, I forgot to answer this.

I think this should be possible to be done, let me try this.

**Review by vfdev-5 — APPROVED** [review]
LGTM !

**Pull request merged**